keps/sig-node/127-usernamespaces-support/README.md

```diff
@@ -0,0 +1,630 @@
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+Container security consists of many different kernel features that work
+together to make containers secure. User namespaces is one such features that
+enables interesting possibilities for containers by allowing them to be root
+inside the container while not being root on the host. This gives more
+capabilities to the containers while protecting the host from the container
+being root and adds one more layer to container security.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+From [user_namespaces(7)](https://man7.org/linux/man-pages/man7/user_namespaces.7.html):
+> User namespaces isolate security-related identifiers and attributes, in
+particular, user IDs and group IDs, the root directory, keys, and capabilities.
+A process's user and group IDs can be different inside and outside a user
+namespace. In particular, a process can have a normal unprivileged user ID
+outside a user namespace while at the same time having a user ID of 0 inside
+the namespace; in other words, the process has full privileges for operations
+inside the user namespace, but is unprivileged for operations outside the
+namespace.
+
+Linux user namespaces allows to run Pods with software that expects to run as
+root or with elevated privileges while still restricting the processes
+permissions on the node, as the effective user ID on the node can be
+unprivileged. This protects the node and other pods running in the node.
+
+The purpose of using user namespaces in Kubernetes is to let the processes in
+Pods think they run as one uid set when in fact they run as different effective
+uids on the host. If a process is able to break into the host, it'll have limited
+impact as it'll be running as an unprivileged user.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- Increase node to Pod isolation in Kubernetes by mapping the root user in a
+Pod to an unprivileged user in the node.
+- Make workloads that require root user inside the container safer for the
+node using the additional security of user namespaces.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+- To have different uid/gid mappings per pod/container. All pods/containers
```

Comment:
Can we have another KEP for this in future? Or it won't be supported? (why?)

Reply:
Definitely. Our plan is to scope this KEP to some specific use cases now and leave further improvements like that for the future.